File: react-native/src/components/common/Ring.js
/*
 * Copyright (C) 2016-2019 Álinson Santos Xavier <isoron@gmail.com>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

import React from 'react';
import PropTypes from 'prop-types';
import Svg, { Circle } from 'react-native-svg';

export default function Ring(props) {
  const {
    size, strokeWidth, color, percentage,
  } = props;

  return (
    <Svg
      height={size}
      width={size}
      viewBox="0 0 100 100"
      style={{ transform: [{ rotate: '-90deg' }] }}
    >
      <Circle
        cx="50"
        cy="50"
        r={50 - strokeWidth / 2}
        stroke="#eee"
        strokeWidth={strokeWidth}
        fill="transparent"
      />
      <Circle
        cx="50"
        cy="50"
        r={50 - strokeWidth / 2}
        stroke={color}
        strokeWidth={strokeWidth}
        fill="transparent"
        strokeDashoffset={`${2 * Math.PI * 45 * (1.0 - percentage)}`}
        strokeDasharray={`${2 * Math.PI * 45}`}
      />
    </Svg>
  );
}

Ring.propTypes = {
  size: PropTypes.number.isRequired,
  strokeWidth: PropTypes.number.isRequired,
  color: PropTypes.string.isRequired,
  percentage: PropTypes.number.isRequired,
};

File: uhabits-core/src/jvmMain/java/org/isoron/uhabits/core/ui/views/Ring.kt
/*
 * Copyright (C) 2016-2021 Álinson Santos Xavier <git@axavier.org>
 *
 * This file is part of Loop Habit Tracker.
 *
 * Loop Habit Tracker is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by the
 * Free Software Foundation, either version 3 of the License, or (at your
 * option) any later version.
 *
 * Loop Habit Tracker is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License for
 * more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program. If not, see <http://www.gnu.org/licenses/>.
 */

package org.isoron.uhabits.core.ui.views

import org.isoron.platform.gui.Canvas
import org.isoron.platform.gui.Color
import org.isoron.platform.gui.View
import java.lang.String.format
import kotlin.math.max
import kotlin.math.min

class Ring(
    val color: Color,
    val percentage: Double,
    val thickness: Double,
    val radius: Double,
    val theme: Theme,
    val label: Boolean = false
) : View {

    override fun draw(canvas: Canvas) {
        val width = canvas.getWidth()
        val height = canvas.getHeight()
        val angle = 360.0 * max(0.0, min(360.0, percentage))

        canvas.setColor(theme.lowContrastTextColor)
        canvas.fillCircle(width / 2, height / 2, radius)

        canvas.setColor(color)
        canvas.fillArc(width / 2, height / 2, radius, 90.0, -angle)

        canvas.setColor(theme.cardBackgroundColor)
        canvas.fillCircle(width / 2, height / 2, radius - thickness)

        if (label) {
            canvas.setColor(color)
            canvas.setFontSize(radius * 0.4)
            canvas.drawText(format("%.0f%%", percentage * 100), width / 2, height / 2)
        }
    }
}
